File: Data Structures & Algorithms/is-palindrome/submission-0.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        i , j = 0, len(s) - 1

        while i < j:
            
            if not s[j].isalnum():
                j -= 1
                continue
            if not s[i].isalnum():
                i += 1 
                continue
            if s[j].lower() == s[i].lower():
                j -= 1
                i +=1 
            else:
                return False
        return True

File: Data Structures & Algorithms/palindrome-partitioning/submission-0.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        vector<string> path;
        party(0, path, res, s);

        return res;
    }

    bool isPalindrome(string s)
    {
        int left = 0, right = s.size() - 1;

        while(left <= right)
        {
            if(s[left] != s[right])
                return false;
            left++;
            right--;
        }

        return true;
    }
    

    void party(int start, vector<string>& path, vector<vector<string>>& res, string& origin)
    {
        if(start == origin.size())
        {
            res.push_back(path);
        }

        for(int end = start; end < origin.size(); end++)
        {
            string subStr = origin.substr(start, end - start + 1);

            if(isPalindrome(subStr))
            {
                path.push_back(subStr);
                party(end + 1, path, res, origin);
                path.pop_back();
            }
        }
    }
};
